src/main/scala/workflow/WorkflowUtils.scala

```diff
@@ -159,4 +159,99 @@ object WorkflowUtils {
       case (id, partition) => Iterator.single((id, partition.length))
     }.collect().toMap
   }
+
+  /**
+   * Remove the given set of instruction indices from the pipeline, updating all dependency pointers
+   * appropriately. Errors when the removed instructions have dependencies on them
+   *
+   * @param setToRemove
+   * @param instructions
+   * @return A tuple containing the new instructions, and
+   *         a mapping of old dependency index to new dependency index
+   */
+  def removeInstructions(setToRemove: Set[Int], instructions: Seq[Instruction]): (Seq[Instruction], Int => Int) = {
+    val offsets = collection.mutable.Map[Int, Int]()
+    for (indexToRemove <- setToRemove;
+         indexToOffset <- (indexToRemove + 1) until instructions.length) {
+      offsets(indexToOffset) = offsets.getOrElse(indexToOffset, 0) + 1
+    }
+
+    val newInstructions = instructions.zipWithIndex.collect {
+      case (instruction, i) if !setToRemove.contains(i) =>
+        instruction.mapDependencies(oldDep =>
+          if (setToRemove.contains(oldDep)) {
+            throw new RuntimeException("WorkflowUtils.removeInstructions attempted to break a dependency")
+          } else {
+            oldDep - offsets.getOrElse(oldDep, 0)
+          }
+        )
+    }
+
+    val oldToNewIndexMapping = (i: Int) => i - offsets.getOrElse(i, 0)
+
+    (newInstructions, oldToNewIndexMapping)
+  }
+
+  def disconnectAndRemoveInstructions(dependencyReplacement: Map[Int, Int], instructions: Seq[Instruction])
```

Comment:
Can we add some scaladoc explaining what the arguments are and what this is doing? This is a pretty heavily used procedure.